executor/join.go

```diff
@@ -286,34 +302,108 @@ func (e *HashJoinExec) fetchInnerRows(ctx context.Context, chkCh chan<- *chunk.C
 	}
 }
 
-// evalRadixBitNum evaluates the radix bit numbers.
-func (e *HashJoinExec) evalRadixBitNum() {
+// partitionInnerRows re-order e.innerResults into sub-relations.
+func (e *HashJoinExec) partitionInnerRows() error {
+	if err := e.preAlloc4InnerParts(); err != nil {
+		return err
+	}
+
+	wg := &sync.WaitGroup{}
+	defer wg.Wait()
+	for i := 0; i < e.partConcurrency; i++ {
+		wg.Add(1)
+		workerID := i
+		go util.WithRecovery(func() {
+			e.doInnerPartition(workerID, wg)
+		}, nil)
+	}
+	return nil
+}
+
+func (e *HashJoinExec) doInnerPartition(startIdx int, wg *sync.WaitGroup) {
+	defer wg.Done()
+	chkIdx, chkNum := startIdx, e.innerResult.NumChunks()
+	for ; chkIdx < chkNum; chkIdx += e.partConcurrency {
+		chk := e.innerResult.GetChunk(chkIdx)
+		for rowIdx, partPtr := range e.innerRowPrts[chkIdx] {
+			partIdx, rowPtr := partPtr.partitionIdx, partPtr.RowPtr
+			part := e.innerParts[partIdx]
+			part.Insert(rowPtr, chk.GetRow(rowIdx))
```

Comment:
```go
part.Insert(rowPtr, e.innerResult.GetRow(rowPtr))
```

Reply:
It's not an equivalent transformation。
`rowPtr` refers to the position in a specific partition.
`rowIdx` refers to the position in current chk of `e.innerResults`.

Reply:
got it.